Implement torch.kaiser_window

muthuArivoli opened the pull request:
Related to #38349

muthuArivoli committed:
initial implementation

muthuArivoli committed:
add kernels

muthuArivoli committed:
fix cuda

muthuArivoli committed:
add headers

muthuArivoli committed:
fix docs and attempt to fix xla

muthuArivoli commented:
@mruberry PTAL. The XLA failure is real, but I'm not sure why it works for the other window functions. but fails on kaiser_window (maybe because I'm using a kernel for kaiser).

In response to the comments from the i0 PR:

1. I believe I have fused the operations (not completely sure if I did it correctly, though). I placed the kernel in the UnaryOpsKernel files, since there was no file in the cpu directory for TensorFactories, but I can create one if that is needed.

2. This function currently uses the periodic arg, which acts opposite to the sym arg in scipy.signal. I did this because all of the other pytorch window functions use this periodic arg, and do not use the sym version.

mruberry commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -395,6 +396,15 @@ static void clamp_min_kernel(TensorIterator& iter, Scalar min_scalar) {
+  AT_DISPATCH_FLOATING_TYPES_AND(kBFloat16, iter.dtype(), "kaiser_window_cpu", [&](){
+    auto alpha = (window_length - 1) / 2.0;
+    cpu_kernel(iter, [=](scalar_t a){
+        return static_cast<scalar_t>(calc_i0(beta * std::sqrt(1 - std::pow((a - alpha) / alpha, 2.0))) / calc_i0(beta));
```
Can you elaborate on this formulation of the window?

The NumPy documentation of the window appears to be incorrect. The NumPy implementation is:

https://github.com/numpy/numpy/blob/54811204a63c2606e89c4f053c4b9a7215207079/numpy/lib/function_base.py#L3328

and the documentation is [here](https://numpy.org/doc/stable/reference/generated/numpy.kaiser.html?highlight=kaiser#numpy.kaiser).

This follows the NumPy implementation and the [Mathworks documentation](https://www.mathworks.com/help/signal/ref/kaiser.html), which is equivalent to [Wikipedia's description](https://en.wikipedia.org/wiki/Kaiser_window) when "symmetrically sampling" the window. 

In this case `a` is a positive integer and `alpha` is always a positive float >= 1, is it actually preferable to divide `a - alpha` by `alpha` or just to divide `a` by `alpha` and subtract one afterwards?

muthuArivoli replied:
Yes, you are right, the numpy and scipy documentation for this function is wrong. I copied their function documentation, but I'll change it to the Mathworks documentation.

I can change it to divide first, then subtract one. I'm not sure if it will really make a difference.

muthuArivoli replied:
I am sorry, I was mistaken - their documentation is actually correct. You actually found the key point in a later comment, with the range that their n is defined in. Numpy shifted the the range leftwards by the N/2 term in the mathworks docs, so that the range for n is -N/2 to N/2, while in mathworks it is 0 to N. In the numerator in mathworks, they do n-N/2, but since numpy has already shifted their range, they do not need to subtract anything. The extra 4 in the numerator in numpy is from the the mathworks denominator - they square (N/2) in the denominator, while numpy only squares N in the denominator.

Let me know which formula is preferred for the docs.

mruberry replied:
Oh that's really interesting. Thank you for explaining this discrepancy to me. As mentioned below I really struggled with the way NumPy was presenting the range of the value. I think taking a couple points from your explanation here would be helpful. I would stick with what makes the most sense to you.

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -982,6 +982,34 @@ Tensor hann_window(
+    window_length += 1;
+  }
+  auto initial = native::arange(window_length, options);
+  auto window = native::zeros(window_length, options);
```
Why `zeros` and not `empty`? Is it because there's a dispatch issue with `native::empty`?

mruberry replied:
Follow-up: seems like we should just use `at::empty` here

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -982,6 +982,34 @@ Tensor hann_window(
+  auto initial = native::arange(window_length, options);
+  auto window = native::zeros(window_length, options);
+  auto iter = TensorIterator::unary_op(window, initial);
+  kaiser_window_stub(iter.device_type(), iter, window_length, beta);
```
Writing a custom kernel to fuse the operations seems like a performance win. Are there any other reasons you decided to write a custom kernel (vs implementing the window as a composite of existing Torch operations)?

muthuArivoli replied:
No, the main/only reason was for performance, as mentioned in the i0 PR.

mruberry commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -395,6 +396,15 @@ static void clamp_min_kernel(TensorIterator& iter, Scalar min_scalar) {
 
+static void kaiser_window_kernel(TensorIterator& iter, int64_t window_length, double beta){
+  AT_DISPATCH_FLOATING_TYPES_AND(kBFloat16, iter.dtype(), "kaiser_window_cpu", [&](){
+    auto alpha = (window_length - 1) / 2.0;
```
This is a little misleading since it will always be double. That means the computation will always occur in double, too. Is that what's intended?

muthuArivoli replied:
No, I forgot to cast, but I'll be fixing this.

mruberry commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -183,6 +184,17 @@ void clamp_max_kernel_cuda(TensorIterator& iter, Scalar max_value) {
+void kaiser_window_kernel_cuda(TensorIterator& iter, int64_t window_length, double beta){
+  AT_DISPATCH_FLOATING_TYPES_AND2(ScalarType::Half, ScalarType::BFloat16, iter.dtype(), "kaiser_window_cuda", [&](){
+    AT_SKIP_BFLOAT16_IF_NOT_ROCM(scalar_t, "kaiser_window_cuda", [&] {
+      auto alpha = (window_length - 1) / 2.0;
```
As with the C++ kernel, this will be double and it will make the entire computation occur in double. Is that intended? Seems like it should compute in the type it's given? (Yes, the `i0` function will internally compute in the accumulate type.)

mruberry commented on test/test_torch.py:
```diff
@@ -8053,28 +8053,30 @@ def test_complex_rot90(self, device, dtype):
+                    res = torch_method(size, *args, periodic=periodic, device=device)
                     # NB: scipy always returns a float32 result
-                    ref = torch.from_numpy(signal.get_window(name, size, fftbins=periodic))
+                    ref = torch.from_numpy(signal.get_window((name, *args), size, fftbins=periodic))
```
I don't think the signature for `get_window()` will do what you like. Instead you can pass the torch function and the reference as a pair:

```
(torch.kaiser_window, scipy.signal.kaiser)
```

mruberry commented on test/test_torch.py:
```diff
@@ -8053,28 +8053,30 @@ def test_complex_rot90(self, device, dtype):
             np_fn = partial(np.rot90, k=rot_times, axes=[0, 1])
             self.compare_with_numpy(torch_fn, np_fn, data)
 
+    @unittest.skipIf(not TEST_SCIPY, "Scipy not found")
```
Does the OSS CI have SciPy? This test could also use NumPy's `kaiser_window`.

muthuArivoli replied:
I believe that it does have Scipy, the tests don't get skipped, from what I can see. I won't be able to test periodic behavior with numpy (only symmetric), so I'll probably stick with scipy.

mruberry commented on test/test_torch.py:
```diff
@@ -8053,28 +8053,30 @@ def test_complex_rot90(self, device, dtype):
-            raise unittest.SkipTest('Scipy not found')
 
-        def test(name):
+        def test(name, args=()):
```
Are the degenerate cases (window size 0 and 1) tested anywhere?

mruberry commented on test/test_torch.py:
```diff
@@ -8053,28 +8053,30 @@ def test_complex_rot90(self, device, dtype):
             test(window)
 
+        for num_test in range(50):
+            test('kaiser', (random.random() * 30,))
```
It'd be nice to see a bigger cross of window sizes and beta params, that can be crossed with the periodic argument, too.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8928,6 +8928,50 @@ def merge_dicts(*dicts):
+kaiser_window(window_length, beta, periodic=True, dtype=None, \
+layout=torch.strided, device=None, requires_grad=False) -> Tensor
+""" + r"""
+Kaiser window function.
```
Would a more accurate description be something like, "Computes symmetric samples from the Kaiser windowing function"?

muthuArivoli replied:
I don't think I agree with this, since you can compute non symmetric samples with this function (in fact the default is to do that).

mruberry replied:
That makes sense, I realize now what I suggested is a problem. What I'm trying to get at is something more like, "samples symmetrically from..." as in this [Wikipedia article](https://en.wikipedia.org/wiki/Kaiser_window#:~:text=It%20is%20a%20one%2Dparameter,which%20is%20difficult%20to%20compute.). But all the other documentation, like NumPy and matlab, seem to just refer to this as computing the Kaiser window function, which doesn't seem correct but is probably good enough for now.

What about just, "Computes the Kaiser window with window length :attr:\`window_length\` and shape parameter :attr:\`beta\`."? It's nice to include how the positional arguments relate to the function in the opening sentence. I don't love the user of the term "shape factor" for beta, but it seems reasonable and consistent with other descriptions.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8928,6 +8928,50 @@ def merge_dicts(*dicts):
 
 
+add_docstr(torch.kaiser_window,
+           """
```
I don't think you need a newline for this triple quote

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8928,6 +8928,50 @@ def merge_dicts(*dicts):
 
+add_docstr(torch.kaiser_window,
+           """
+kaiser_window(window_length, beta, periodic=True, dtype=None, \
```
`*` should be in here somewhere

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8928,6 +8928,50 @@ def merge_dicts(*dicts):
+Kaiser window function.
+
+.. math::
+    w[n] = I_0 \left( \beta \sqrt{1 - \frac{4n^2}{(M-1)^2}} \right) / I_0( \beta )
```
See note about documenting the function above.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8928,6 +8928,50 @@ def merge_dicts(*dicts):
+with
+
+.. math::
+    -\frac{M-1}{2} \leq n \leq \frac{M-1}{2}
```
I don't understand this bit in the original NumPy documentation. What is this trying to say?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8928,6 +8928,50 @@ def merge_dicts(*dicts):
+    -\frac{M-1}{2} \leq n \leq \frac{M-1}{2}
+
+where :math:`N` is the full window size.
+The input :attr:`window_length` is a positive integer controlling the
```
newline

mruberry replied:
This seems appropriate for the description of `window_length` under args.

mruberry commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -183,6 +184,17 @@ void clamp_max_kernel_cuda(TensorIterator& iter, Scalar max_value) {
+    AT_SKIP_BFLOAT16_IF_NOT_ROCM(scalar_t, "kaiser_window_cuda", [&] {
+      auto alpha = (window_length - 1) / 2.0;
+      gpu_kernel(iter, [=]GPU_LAMBDA(scalar_t a) -> scalar_t {
+        return static_cast<scalar_t>(calc_i0(beta * ::sqrt(1 - ::pow((a - alpha) / alpha, 2.0))) / calc_i0(beta));
```
TODO: try precomputing and passing `calc_i0(beta)` as a parameter?

mruberry replied:
(TODO for another PR, I mean)

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8928,6 +8928,50 @@ def merge_dicts(*dicts):
+
+where :math:`N` is the full window size.
+The input :attr:`window_length` is a positive integer controlling the
+returned window size. :attr:`periodic` flag determines whether the returned
```
I like the discussion of the periodic flag here, but it needs to be clearer. Would it be simpler to say that setting periodic to True is equivalent to computing as if the window length was L + 1? Then you can explain why this option is available.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8928,6 +8928,50 @@ def merge_dicts(*dicts):
+``torch.kaiser_window(L + 1, B, periodic=False)[:-1])``.
+
+.. note::
+    If :attr:`window_length` :math:`=1`, the returned window contains a single value 1.
```
newline here. This is actually interesting that this is true with periodic being True, too. What do you think about that interaction?

muthuArivoli replied:
All the other window functions in pytorch, and those in scipy/numpy also follow this behavior. I feel like any other value would simply not be correct, since I would expect the value to be representative of the window, and with one being located at the center of the window, I would think that for a window length of 1, that one would be what is returned.

mruberry replied:
Sounds good.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8928,6 +8928,50 @@ def merge_dicts(*dicts):
+""" + r"""
+Arguments:
+    window_length (int): the size of returned window
+    periodic (bool, optional): If True, returns a window to be used as periodic
```
This should describe exactly what periodic does before describing what it's intended for.

mruberry replied:
"Keyword args:" goes in here somewhere

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8928,6 +8928,50 @@ def merge_dicts(*dicts):
+    window_length (int): the size of returned window
+    periodic (bool, optional): If True, returns a window to be used as periodic
+        function. If False, return a symmetric window.
+    {dtype} Only floating point types are supported.
```
This will automatically generate the description for dtype, so you don't need to add one.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8928,6 +8928,50 @@ def merge_dicts(*dicts):
+    {requires_grad}
+
+Returns:
+    Tensor: A 1-D tensor of size :math:`(\text{{window\_length}},)` containing the window
```
More accurately it contains samples from the window, right?

mruberry commented:
Hey @muthuArivoli! This PR looks pretty good. I have some questions about the documentation of the windowing function, and some suggestions for how to extend the tests. I wonder if this shouldn't just add a new test just for the kaiser window where you can test a wider variety of lengths, beta terms, and true/false values for the period kwarg. Also, although not mentioned in the comments above, the TensorOption arguments, like dtype, need to be tested, too. 

I'm not sure what's causing the XLA failure but you shouldn't expect XLA to work with this PR because you've written your own CPU and CUDA kernels. XLA can't understand those. That's OK - you can just skip the test on XLA (@onlyOnCPUAndCUDA).

mruberry commented:
> @mruberry PTAL. The XLA failure is real, but I'm not sure why it works for the other window functions. but fails on kaiser_window (maybe because I'm using a kernel for kaiser).

See my other comment but we should just skip the test on XLA. It won't understand how to run this op.

> In response to the comments from the i0 PR:
> 
> 1. I believe I have fused the operations (not completely sure if I did it correctly, though). I placed the kernel in the UnaryOpsKernel files, since there was no file in the cpu directory for TensorFactories, but I can create one if that is needed.

No, that's fine. 

> 2. This function currently uses the periodic arg, which acts opposite to the sym arg in scipy.signal. I did this because all of the other pytorch window functions use this periodic arg, and do not use the sym version.

That makes sense. Thank you.

muthuArivoli committed:
fix kernels

muthuArivoli committed:
update docs

muthuArivoli committed:
add zero window size to test

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-kaiser-window

muthuArivoli commented:
@mruberry I have updated with most of the suggestions. Could use a review now, especially to confirm if the casting is right in the kernels. A few notes:

I have tried updating the docs to make them more clear, but I'm not sure if I made them worse. With regards to kwargs, I have made the periodic arg also positional, is that desired?

I want a little bit more guidance on testing. I added 0 as a test case for window size. Currently, I test against 50 different random beta values, each with 2 special window sizes (0 and 1) and 7 regular window sizes, and against both symmetric and periodic for all of these combinations. How many more lengths/window sizes combinations do you think is needed for testing?

Finally, the asan test is failing for the blackman window - I guessing this is due to me adding the zero window size as a test case. I'll look into it further.

mruberry commented:
> Finally, the asan test is failing for the blackman window - I guessing this is due to me adding the zero window size as a test case. I'll look into it further.

Blackman needs a special case for a window length of zero because of this line:

https://github.com/pytorch/pytorch/blob/c88ac2567935c37780e21c65ab4c43ea9f817373/aten/src/ATen/native/TensorFactories.cpp#L921

Just before this is:

```
if (periodic) {
    window_length += 1;
  }
```

which is what's causing the division by zero error.

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -982,6 +982,34 @@ Tensor hann_window(
+  return native::kaiser_window(window_length, beta, /*periodic=*/true, options);
+}
+
+Tensor kaiser_window(
```
To be consistent with PyTorch's windowing functions the name is `kaiser_window`, like you have here, but the order of arguments would be `window_length, periodic, beta`. See, for example, the Hamming window documentation which has alpha and beta terms:

https://pytorch.org/docs/master/generated/torch.hamming_window.html?highlight=window#torch.hamming_window

To be consistent with NumPy we can add a `kaiser` function later that has a NumPy-compatible signature.

mruberry commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -395,6 +396,15 @@ static void clamp_min_kernel(TensorIterator& iter, Scalar min_scalar) {
 
+static void kaiser_window_kernel(TensorIterator& iter, int64_t window_length, double beta){
+  AT_DISPATCH_FLOATING_TYPES_AND(kBFloat16, iter.dtype(), "kaiser_window_cpu", [&](){
+    auto alpha = static_cast<scalar_t>((window_length - 1) / 2.0);
```
This looks correct now. As a nit you could do `const scalar_t alpha = ...` to be more explicit.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1646,6 +1646,12 @@
 - func: hamming_window.periodic_alpha_beta(int window_length, bool periodic, float alpha, float beta, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
   use_c10_dispatcher: full
 
+- func: kaiser_window(int window_length, float beta, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
See note above about how we should derive arg order. I actually think PyTorch's ordering is a little silly, and agree it's more natural to put arguments like `beta` before arguments like `periodic`. However, we can fix that when we implement the NumPy names (that is, `kaiser` instead of `kaiser_window`). The one caveat with that plan is that NumPy's name for the Hann window is incorrect, but we'll get to that later (I filed an issue with NumPy for this https://github.com/numpy/numpy/issues/17303). 

This suggestion requires setting a default beta parameter. matlab uses 1/2, the NumPy documentation suggests 14, and TensorFlow's default is 12. 12 seems reasonable to me.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8930,6 +8930,54 @@ def merge_dicts(*dicts):
+Kaiser window function.
+
+.. math::
+    w[n] = I_0 \left( \beta \sqrt{1 - \frac{4n^2}{(M-1)^2}} \right) / I_0( \beta )
```
Can we rewrite this in terms of the inputs, which don't involve `n` or `M`?

```
w_i = ...
```

and then ditch the inequality on `n` below and the definition of `M`? If we do want to define `M` for clarity let's define it BEFORE it's used in the expression.

mruberry replied:
See the documentation for `torch.sub` for an example using of the mathematical expression using underscore notation:

https://pytorch.org/docs/master/generated/torch.sub.html?highlight=sub#torch.sub

mruberry replied:
Speaking of defining things before they're used, I know we know what `i0` is (because you just made it), but this should define what `i0` and link to its documentation.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8930,6 +8930,54 @@ def merge_dicts(*dicts):
+
+where :math:`M` is the full window size.
+
+If :attr:`periodic` is true, the :math:`M` in
```
This is a tricky paragraph. Let me suggest another idea. If `M` is defined in advance we can say when it's defined:

"Letting i0 be the zeroth order modified Bessel function of the first kind (see torch.i0) and M = :attr:\'window_length\' if :attr:\'periodic\' is False and :attr:\`window_length\` + 1 if :attr:\`periodic\` is True, this computes:"

Then the mathematical expression as dicussed above. 

Then after the mathematical expression we have this paragraph, which can elaborate on the intent of the periodic argument. Maybe something like:

"Calling torch.kaiser_window(L, B, periodic=True) is equivalent to torch.kaiser_window(L + 1, B, periodic=False)[:-1]). The :attr:\`periodic\` argument is intended as a helpful shorthand to produce a periodic window as input to functions like :func:\`torch.stft\`."

Note the use of :func: and not :meth for the torch.stft reference. 

Does that split sound more reasonable?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8930,6 +8930,54 @@ def merge_dicts(*dicts):
+Args:
+    window_length (int):  positive integer controlling the returned window size.
+    beta: shape parameter for the window
+    periodic (bool, optional): If True, returns a periodic window, for use in spectral analysis.
```
"returns a periodic window suitable for use in..."

and 

"returns a symmetric window suitable for use in filter design"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8930,6 +8930,54 @@ def merge_dicts(*dicts):
+    {requires_grad}
+
+Returns:
+    Tensor: A 1-D tensor of size :math:`(\text{{window\_length}},)` containing samples from the window
```
I don't think you need a "Returns:" here. The return type is clear. Typically we only include this if the return type is complicated or unexpected. For example, if the returned value is a tuple of two tensors.

mruberry commented:
> @mruberry I have updated with most of the suggestions. Could use a review now, especially to confirm if the casting is right in the kernels. A few notes:

The casts in the kernels look good.

> I have tried updating the docs to make them more clear, but I'm not sure if I made them worse. With regards to kwargs, I have made the periodic arg also positional, is that desired?

I think we're getting close with the docs. I added some comments and references about arg position and new doc suggestions.

> I want a little bit more guidance on testing. I added 0 as a test case for window size. Currently, I test against 50 different random beta values, each with 2 special window sizes (0 and 1) and 7 regular window sizes, and against both symmetric and periodic for all of these combinations. How many more lengths/window sizes combinations do you think is needed for testing?

Testing against 50 random beta values is very thorough. These combinations sound sufficient. In the future we should test with the tensor options arguments, like dtype and layout, too, but that's separable and shouldn't be in this PR.

muthuArivoli committed:
review fixes

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-kaiser-window

muthuArivoli committed:
fix lint and test

muthuArivoli committed:
fix tests and docs

muthuArivoli commented:
@mruberry, Unfortunately, the doc build keeps failing if I list out the keyword args, giving me the same error as in #43669. I also tried the fix recommended in that issue, and it still failed to build. I have removed the keyword args description, but kept the star in the function description. What is the right way to proceed?

mruberry commented:
> @mruberry, Unfortunately, the doc build keeps failing if I list out the keyword args, giving me the same error as in #43669. I also tried the fix recommended in that issue, and it still failed to build. I have removed the keyword args description, but kept the star in the function description. What is the right way to proceed?

That's the best we can do for now. Sorry about the docs build. We should really root cause and fix this issue.

muthuArivoli commented:
> > @mruberry, Unfortunately, the doc build keeps failing if I list out the keyword args, giving me the same error as in #43669. I also tried the fix recommended in that issue, and it still failed to build. I have removed the keyword args description, but kept the star in the function description. What is the right way to proceed?
> 
> That's the best we can do for now. Sorry about the docs build. We should really root cause and fix this issue.

Alright, in that case I think this can use another review, hopefully the content in the docs is better now.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1656,6 +1656,15 @@
+- func: kaiser_window(int window_length, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
+  use_c10_dispatcher: full
+
+- func: kaiser_window.periodic(int window_length, bool periodic, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
Can these three native_functions.yaml entries be combined if defaults for periodic and beta are specified?

```
- func: kaiser_window.beta(int window_length, bool periodic=True, float beta=12, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
  use_c10_dispatcher: full
```

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8977,6 +8977,42 @@ def merge_dicts(*dicts):
+to produce a periodic window as input to functions like :func:`torch.stft`.
+
+.. note::
+    If :attr:`window_length` :math:`=1`, the returned window contains a single value 1.
```
Nit: "If :attr:\`window_length\` is one then the returned window is a single element tensor containing a one."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8977,6 +8977,42 @@ def merge_dicts(*dicts):
+
+""" + r"""
+Args:
+    window_length (int):  positive integer controlling the returned window size.
```
just say "length of the window"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8977,6 +8977,42 @@ def merge_dicts(*dicts):
+    window_length (int):  positive integer controlling the returned window size.
+    periodic (bool, optional): If True, returns a periodic window suitable for use in spectral analysis.
+        If False, returns a symmetric window suitable for use in filter design.
+    beta (float, optional): shape parameter for the window
```
nit: "window" -> "window."

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -914,6 +914,9 @@ Tensor blackman_window(
     bool periodic,
     const TensorOptions& options) {
   window_function_checks("blackman_window", options, window_length);
+  if (window_length == 0) {
```
Thank you for fixing this!

mruberry reviewed (approved):
Great work, @muthuArivoli! i0 and this are really interesting functions and it's great to have them available in PyTorch. 

I have a couple nits about the docs. I also think you could get away with a single dispatch for this function, too, but it's OK if you'd prefer to leave the three dispatch entries you have (that's consistent with the other window functions, after all). Let me know when you're satisfied and we'll get this merged!

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-kaiser-window

muthuArivoli committed:
update docs

muthuArivoli commented:
@mruberry, I updated the docs. I tried implementing the single dispatch, but I ran into this compile error:
```
../torch/csrc/autograd/generated/variable_factories.h:371:72: error: ‘True’ was not declared in this scope
 inline at::Tensor kaiser_window(int64_t window_length, bool periodic = True, double beta = 12.0, const at::TensorOptions & options = {}) {
                                                                        ^~~~
```
I tried to make the True lowercase in the yaml file, but this still popped up. If this is due to something not related to the implementation of the kaiser window function, then I think I'll stick with the multiple dispatches, and this should be good to merge.

mruberry commented:
> @mruberry, I updated the docs. I tried implementing the single dispatch, but I ran into this compile error:
> 
> ```
> ../torch/csrc/autograd/generated/variable_factories.h:371:72: error: ‘True’ was not declared in this scope
>  inline at::Tensor kaiser_window(int64_t window_length, bool periodic = True, double beta = 12.0, const at::TensorOptions & options = {}) {
>                                                                         ^~~~
> ```
> 
> I tried to make the True lowercase in the yaml file, but this still popped up. If this is due to something not related to the implementation of the kaiser window function, then I think I'll stick with the multiple dispatches, and this should be good to merge.

That's weird. See, for example:

https://github.com/pytorch/pytorch/blob/69839ea3f6680d068db38795afe03ae4b8a30962/aten/src/ATen/native/native_functions.yaml#L97

which sets a bool to a default value. We don't need to worry about it in this PR, however. Let's wait for the tests and get it merged.

mruberry commented:
This is merged! Congrats @muthuArivoli. Let me know if you're looking for something else and, if you are, what might be fun.

muthuArivoli commented:
Yes, I'm looking for something else. Any recommendations?

mruberry commented:
It depends on what you're looking for. Let me offer a few options:

- `torch.divmod` would be nice. I think contributors have avoided it because it returns two tensors instead of one. 
- We're starting to invest more in linear algebra operations, like improving our Cholesky decomposition. These are especially challenging tasks but one of them might be interesting.
- [`kron`](https://numpy.org/doc/1.18/reference/generated/numpy.kron.html?highlight=kron#numpy.kron) would be a tricky tensor factory to implement.
- [`pad`](https://numpy.org/doc/stable/reference/generated/numpy.pad.html?highlight=pad#numpy.pad) would be useful. PyTorch already has a `pad` function but there are discrepancies between our Python and C++ APIs, and NumPy supports many more modes (making pad consistent and/or support even one new mode would be nice).

Would any of those be interesting? Or would you like something simpler as a break? I can break down linear algebra-related ideas further if that direction is interesting.

muthuArivoli commented:
I'd like to hear more about the linear algebra ideas.

One other question. When implementing the specialized splits (hsplit, dsplit, vsplit), I ran into a problem where the arg name for the input didn't match up with the arg name in the torch.split function. For some reason the torch.split uses the arg name `tensor` as its input, but I could only get the arg name `input` to work. Any ideas on how to fix that/is that important?

mruberry commented:
> I'd like to hear more about the linear algebra ideas.

I'll pull a few and update the linalg rollup issue soon. It's a really interesting space.

> One other question. When implementing the specialized splits (hsplit, dsplit, vsplit), I ran into a problem where the arg name for the input didn't match up with the arg name in the torch.split function. For some reason the torch.split uses the arg name `tensor` as its input, but I could only get the arg name `input` to work. Any ideas on how to fix that/is that important?

That is strange. I thought we required native_functions always start with `self`, but I might be mistaken. Do you have a draft PR I could look at? Also, it seems like the first arg in `torch.split` should be called `input` in the documentation, to be consistent with other functions.
